##
## note that you must run 'make proto' after applying this patch
## and then rebuild winbindd and smbd.
##
Index: libsmb/samlogon_cache.c
===================================================================
--- samba-3.0.6/source/libsmb/samlogon_cache.c	(revision 2082)
+++ samba-3.0.6/source/libsmb/samlogon_cache.c	(working copy)
@@ -106,9 +106,10 @@
 
 /***********************************************************************
  Store a NET_USER_INFO_3 structure in a tdb for later user 
+ username should be in UTF-8 format
 ***********************************************************************/
 
-BOOL netsamlogon_cache_store(TALLOC_CTX *mem_ctx, NET_USER_INFO_3 *user)
+BOOL netsamlogon_cache_store(TALLOC_CTX *mem_ctx, const char * username, NET_USER_INFO_3 *user)
 {
 	TDB_DATA 	data;
         fstring 	keystr;
@@ -130,6 +131,14 @@
 	slprintf(keystr, sizeof(keystr), "%s", sid_string_static(&user_sid));
 
 	DEBUG(10,("netsamlogon_cache_store: SID [%s]\n", keystr));
+	
+	/* only Samba fills in the username, not sure why NT doesn't */
+	/* so we fill it in since winbindd_getpwnam() makes use of it */
+	
+	if ( !user->uni_user_name.buffer ) {
+		init_unistr2( &user->uni_user_name, username, STR_TERMINATE );
+		init_uni_hdr( &user->hdr_user_name, &user->uni_user_name );
+	}
 		
 	/* Prepare data */
 	
Index: nsswitch/winbindd_pam.c
===================================================================
--- samba-3.0.6/source/nsswitch/winbindd_pam.c	(revision 2082)
+++ samba-3.0.6/source/nsswitch/winbindd_pam.c	(working copy)
@@ -283,7 +283,7 @@
 	}
 	
 	if (NT_STATUS_IS_OK(result)) {
-		netsamlogon_cache_store( cli->mem_ctx, &info3 );
+		netsamlogon_cache_store( cli->mem_ctx, name_user, &info3 );
 		wcache_invalidate_samlogon(find_domain_from_name(name_domain), &info3);
 
 		/* Check if the user is in the right group */
@@ -538,7 +538,7 @@
 	}
 
 	if (NT_STATUS_IS_OK(result)) {
-		netsamlogon_cache_store( cli->mem_ctx, &info3 );
+		netsamlogon_cache_store( cli->mem_ctx, name_user, &info3 );
 		wcache_invalidate_samlogon(find_domain_from_name(name_domain), &info3);
 		
 		if (!NT_STATUS_IS_OK(result = check_info3_in_group(mem_ctx, &info3, state->request.data.auth_crap.required_membership_sid))) {
Index: auth/auth_domain.c
===================================================================
--- samba-3.0.6/source/auth/auth_domain.c	(revision 2082)
+++ samba-3.0.6/source/auth/auth_domain.c	(working copy)
@@ -205,7 +205,7 @@
 	} else {
 		nt_status = make_server_info_info3(mem_ctx, user_info->internal_username.str, 
 						   user_info->smb_name.str, domain, server_info, &info3);
-		netsamlogon_cache_store( mem_ctx, &info3 );
+		netsamlogon_cache_store( mem_ctx, user_info->smb_name.str, &info3 );
 	}
 
 #if 0
Index: smbd/password.c
===================================================================
--- samba-3.0.6/source/smbd/password.c	(revision 2069)
+++ samba-3.0.6/source/smbd/password.c	(working copy)
@@ -252,27 +252,30 @@
 	}
 
 	/* Register a home dir service for this user iff
+	
 	   (a) This is not a guest connection,
 	   (b) we have a home directory defined 
-	   If a share exists by this name (autoloaded or not) reuse it so
-	   long as the home directory is the same as the share directory. */
+	   (c) there s not an existing static share by that name
+	   
+	   If a share exists by this name (autoloaded or not) reuse it . */
 
-	if ( (!vuser->guest) && vuser->unix_homedir && *(vuser->unix_homedir)) {
+	vuser->homes_snum = -1;
+
+	if ( (!vuser->guest) && vuser->unix_homedir && *(vuser->unix_homedir)) 
+	{
 		int servicenumber = lp_servicenumber(vuser->user.unix_name);
+
 		if ( servicenumber == -1 ) {
 			DEBUG(3, ("Adding homes service for user '%s' using home directory: '%s'\n", 
 				vuser->user.unix_name, vuser->unix_homedir));
 			vuser->homes_snum = add_home_service(vuser->user.unix_name, 
 						vuser->user.unix_name, vuser->unix_homedir);
-		} else if (strcmp(lp_pathname(servicenumber),vuser->unix_homedir) == 0) {
-			DEBUG(3, ("Reusing homes service for user '%s' using home directory: '%s'\n", 
-				vuser->user.unix_name, vuser->unix_homedir));
-
+		} else {
+			DEBUG(3, ("Using static (or previously created) service for user '%s'; path = '%s'\n", 
+				vuser->user.unix_name, lp_pathname(servicenumber) ));
 			vuser->homes_snum = servicenumber;
 		}
-	} else {
-		vuser->homes_snum = -1;
-	}
+	} 
 	
 	if (srv_is_signing_negotiated() && !vuser->guest && !srv_signing_started()) {
 		/* Try and turn on server signing on the first non-guest sessionsetup. */
Index: param/loadparm.c
===================================================================
--- samba-3.0.6/source/param/loadparm.c	(revision 2069)
+++ samba-3.0.6/source/param/loadparm.c	(working copy)
@@ -2313,6 +2313,10 @@
 	copy_service(ServicePtrs[i], &tservice, NULL);
 	if (name)
 		string_set(&ServicePtrs[i]->szService, name);
+		
+	DEBUG(8,("add_a_service: Creating snum = %d for %s\n", 
+		i, ServicePtrs[i]->szService));
+		
 	return (i);
 }
 
@@ -2356,7 +2356,7 @@
 	ServicePtrs[i]->autoloaded = True;
 
 	DEBUG(3, ("adding home's share [%s] for user '%s' at '%s'\n", pszHomename, 
-	       user, newHomedir));
+	       user, ServicePtrs[i]->szPath ));
 	
 	return (True);
 }
##
## This patch addresses some printing issues in Samba 3.0.6
## BUG ID's: 1464, 1679 (maybe)
##
## To apply:
##    $ tar zxf samba-3.0.6.tar.gz
##    $ cd samba-3.0.6
##    $ patch -p1 < printing_fixes_v1.patch
##
## now build Samba as usual
##

diff -ruBb --exclude-from=samba-cvs/diff.excludes samba-3.0.6-orig/source/include/rpc_spoolss.h samba-3.0.6/source/include/rpc_spoolss.h
--- samba-3.0.6-orig/source/include/rpc_spoolss.h	2004-04-04 01:37:23.000000000 -0600
+++ samba-3.0.6/source/include/rpc_spoolss.h	2004-09-01 10:13:52.180941448 -0500
@@ -376,6 +376,7 @@
 #define PRINTER_ATTRIBUTE_SAMBA			(PRINTER_ATTRIBUTE_RAW_ONLY|\
 						 PRINTER_ATTRIBUTE_SHARED|\
 						 PRINTER_ATTRIBUTE_LOCAL)
+#define PRINTER_ATTRIBUTE_NOT_SAMBA		(PRINTER_ATTRIBUTE_NETWORK)
 
 #define NO_PRIORITY	 0
 #define MAX_PRIORITY	99
diff -ruBb --exclude-from=samba-cvs/diff.excludes samba-3.0.6-orig/source/param/loadparm.c samba-3.0.6/source/param/loadparm.c
--- samba-3.0.6-orig/source/param/loadparm.c	2004-08-19 08:39:12.000000000 -0500
+++ samba-3.0.6/source/param/loadparm.c	2004-09-01 10:13:52.167943424 -0500
@@ -1325,6 +1325,8 @@
 
 		string_set(&sDefault.fstype, FSTYPE_STRING);
 
+		init_printer_values(&sDefault);
+
 		done_init = True;
 	}
 
@@ -3978,9 +3980,6 @@
 	}
 
 	init_iconv();
-#if 0	/* JERRY */
-	init_printer_values(&sDefault);
-#endif
 
 	return (bRetval);
 }
diff -ruBb --exclude-from=samba-cvs/diff.excludes samba-3.0.6-orig/source/printing/notify.c samba-3.0.6/source/printing/notify.c
--- samba-3.0.6-orig/source/printing/notify.c	2004-07-08 12:06:11.000000000 -0500
+++ samba-3.0.6/source/printing/notify.c	2004-09-01 10:13:52.152945704 -0500
@@ -243,18 +243,21 @@
 	 * as they will just cause flickering updates in the client.
 	 */
 
-	if ((num_messages < 100) && (msg->type == JOB_NOTIFY_TYPE) &&
-				(msg->field == JOB_NOTIFY_TOTAL_BYTES || msg->field == JOB_NOTIFY_TOTAL_PAGES)) {
+	if ((num_messages < 100) && (msg->type == JOB_NOTIFY_TYPE) 
+		&& (msg->field == JOB_NOTIFY_TOTAL_BYTES 
+		    || msg->field == JOB_NOTIFY_TOTAL_PAGES )) 
+	{
 
-		for (tmp_ptr = notify_queue_head; tmp_ptr; tmp_ptr = tmp_ptr->next) {
+		for (tmp_ptr = notify_queue_head; tmp_ptr; tmp_ptr = tmp_ptr->next) 
+		{
 			if (tmp_ptr->msg->type == msg->type &&
 					tmp_ptr->msg->field == msg->field &&
 					tmp_ptr->msg->id == msg->id &&
 					tmp_ptr->msg->flags == msg->flags &&
 					strequal(tmp_ptr->msg->printer, msg->printer)) {
 
-				DEBUG(5, ("send_spoolss_notify2_msg: replacing message 0x%02x/0x%02x for printer %s \
-in notify_queue\n", msg->type, msg->field, msg->printer));
+				DEBUG(5,("send_spoolss_notify2_msg: replacing message 0x%02x/0x%02x for "
+					 "printer %s in notify_queue\n", msg->type, msg->field, msg->printer));
 
 				tmp_ptr->msg = msg;
 				return;
diff -ruBb --exclude-from=samba-cvs/diff.excludes samba-3.0.6-orig/source/printing/nt_printing.c samba-3.0.6/source/printing/nt_printing.c
--- samba-3.0.6-orig/source/printing/nt_printing.c	2004-08-19 08:39:12.000000000 -0500
+++ samba-3.0.6/source/printing/nt_printing.c	2004-09-01 10:13:52.161944336 -0500
@@ -3335,7 +3335,8 @@
 			info.parameters);
 
 	/* Samba has to have shared raw drivers. */
-	info.attributes = PRINTER_ATTRIBUTE_SAMBA;
+	info.attributes |= PRINTER_ATTRIBUTE_SAMBA;
+	info.attributes &= ~PRINTER_ATTRIBUTE_NOT_SAMBA;
 
 	/* Restore the stripped strings. */
 	slprintf(info.servername, sizeof(info.servername)-1, "\\\\%s", get_called_name());
diff -ruBb --exclude-from=samba-cvs/diff.excludes samba-3.0.6-orig/source/printing/printing.c samba-3.0.6/source/printing/printing.c
--- samba-3.0.6-orig/source/printing/printing.c	2004-08-19 08:39:12.000000000 -0500
+++ samba-3.0.6/source/printing/printing.c	2004-09-01 10:13:52.155945248 -0500
@@ -442,22 +442,30 @@
 	if (!old_data)
 		new_job = True;
 
-	/* Notify the job name first */
-
-	if (new_job || !strequal(old_data->jobname, new_data->jobname))
-		notify_job_name(snum, jobid, new_data->jobname);
-
 	/* Job attributes that can't be changed.  We only send
 	   notification for these on a new job. */
 
+ 	/* ACHTUNG!  Due to a bug in Samba's spoolss parsing of the 
+ 	   NOTIFY_INFO_DATA buffer, we *have* to send the job submission 
+ 	   time first or else we'll end up with potential alignment 
+ 	   errors.  I don't think the systemtime should be spooled as 
+ 	   a string, but this gets us around that error.   
+ 	   --jerry (i'll feel dirty for this) */
+ 
 	if (new_job) {
 		notify_job_submitted(snum, jobid, new_data->starttime);
 		notify_job_username(snum, jobid, new_data->user);
 	}
 
+	if (new_job || !strequal(old_data->jobname, new_data->jobname))
+		notify_job_name(snum, jobid, new_data->jobname);
+
 	/* Job attributes of a new job or attributes that can be
 	   modified. */
 
+	if (new_job || !strequal(old_data->jobname, new_data->jobname))
+		notify_job_name(snum, jobid, new_data->jobname);
+
 	if (new_job || old_data->status != new_data->status)
 		notify_job_status(snum, jobid, map_to_spoolss_status(new_data->status));
 
@@ -575,24 +583,17 @@
 		return;
 
 	if (!pjob) {
-		DEBUG(5, ("pjob_delete(): we were asked to delete nonexistent job %u\n",
+		DEBUG(5, ("pjob_delete: we were asked to delete nonexistent job %u\n",
 					(unsigned int)jobid));
 		release_print_db(pdb);
 		return;
 	}
 
-	/* Send a notification that a job has been deleted */
-
-	job_status = map_to_spoolss_status(pjob->status);
-
 	/* We must cycle through JOB_STATUS_DELETING and
            JOB_STATUS_DELETED for the port monitor to delete the job
            properly. */
 	
-	job_status |= JOB_STATUS_DELETING;
-	notify_job_status(snum, jobid, job_status);
-	
-	job_status |= JOB_STATUS_DELETED;
+	job_status = JOB_STATUS_DELETING|JOB_STATUS_DELETED;
 	notify_job_status(snum, jobid, job_status);
 
 	/* Remove from printing.tdb */
diff -ruBb --exclude-from=samba-cvs/diff.excludes samba-3.0.6-orig/source/rpc_server/srv_spoolss_nt.c samba-3.0.6/source/rpc_server/srv_spoolss_nt.c
--- samba-3.0.6-orig/source/rpc_server/srv_spoolss_nt.c	2004-08-19 08:39:13.000000000 -0500
+++ samba-3.0.6/source/rpc_server/srv_spoolss_nt.c	2004-09-01 10:13:52.177941904 -0500
@@ -1026,7 +1026,7 @@
 			sending_msg_count++;
 			
 			
-			DEBUG(10,("process_notify2_message: Sending message type [%x] field [%x] for printer [%s]\n",
+			DEBUG(10,("process_notify2_message: Sending message type [0x%x] field [0x%2x] for printer [%s]\n",
 				msg->type, msg->field, p->dev.handlename));
 
 			/* 
@@ -5574,6 +5574,12 @@
 		return getprinterdriver2_level3(servername, architecture, clientmajorversion, snum, buffer, offered, needed);
 	case 6:
 		return getprinterdriver2_level6(servername, architecture, clientmajorversion, snum, buffer, offered, needed);
+#if 0	/* JERRY */
+	case 101: 
+		/* apparently this call is the equivalent of 
+		   EnumPrinterDataEx() for the DsDriver key */
+		break;
+#endif
 	}
 
 	return WERR_UNKNOWN_LEVEL;
@@ -5935,7 +5941,9 @@
 	slprintf(info->printername, sizeof(info->printername)-1, "\\\\%s\\%s",
 		 get_called_name(), p );
 		 
-	info->attributes = PRINTER_ATTRIBUTE_SAMBA;
+	info->attributes |= PRINTER_ATTRIBUTE_SAMBA;
+	info->attributes &= ~PRINTER_ATTRIBUTE_NOT_SAMBA;
+	
 	
 	
 	return True;
--- samba-3.0.6/examples/LDAP/samba.schema.orig	2004-08-19 09:39:13.000000000 -0400
+++ samba-3.0.6/examples/LDAP/samba.schema	2004-08-20 11:53:00.000000000 -0400
@@ -10,6 +10,25 @@
 ## 1.3.6.1.4.1.7165.2.1.x - attributetypes
 ## 1.3.6.1.4.1.7165.2.2.x - objectclasses
 ##
+## ----- READ THIS WHEN ADDING A NEW ATTRIBUTE OR OBJECT CLASS ------
+##
+## Run the 'get_next_oid' bash script in this directory to find the 
+## next available OID for attribute type and object classes.
+##
+##   $ ./get_next_oid
+##   attributetype ( 1.3.6.1.4.1.7165.2.1.XX NAME ....
+##   objectclass ( 1.3.6.1.4.1.7165.2.2.XX NAME ....
+##
+## Also ensure that new entries adhere to the declaration style
+## used throughout this file
+##
+##    <attributetype|objectclass> ( 1.3.6.1.4.1.7165.2.XX.XX NAME ....
+##                               ^ ^                        ^
+##
+## The spaces are required for the get_next_oid script (and for 
+## readability).
+##
+## ------------------------------------------------------------------
 
 ########################################################################
 ##                            HISTORICAL                              ##
@@ -212,7 +231,7 @@
 	EQUALITY integerMatch
 	SYNTAX 1.3.6.1.4.1.1466.115.121.1.27 SINGLE-VALUE )
 
-attributetype ( 1.3.6.1.4.1.7165.2.1.50 NAME 'sambaLogonHours'
+attributetype ( 1.3.6.1.4.1.7165.2.1.55 NAME 'sambaLogonHours'
 	DESC 'Logon Hours'
 	EQUALITY caseIgnoreIA5Match
 	SYNTAX 1.3.6.1.4.1.1466.115.121.1.26{42} SINGLE-VALUE )
@@ -255,7 +274,7 @@
 	EQUALITY caseExactMatch
 	SYNTAX 1.3.6.1.4.1.1466.115.121.1.15{1050} )
 
-attributetype ( 1.3.6.1.4.1.7165.2.1.50 NAME 'sambaPasswordHistory'
+attributetype ( 1.3.6.1.4.1.7165.2.1.54 NAME 'sambaPasswordHistory'
 	DESC 'Concatenated MD4 hashes of the unicode passwords used on this account'
 	EQUALITY caseIgnoreIA5Match
 	SYNTAX 1.3.6.1.4.1.1466.115.121.1.26{32} )
@@ -316,6 +335,51 @@
 	EQUALITY integerMatch
 	SYNTAX 1.3.6.1.4.1.1466.115.121.1.27 SINGLE-VALUE )
 
+attributetype ( 1.3.6.1.4.1.7165.2.1.41 NAME 'sambaShareName'
+	DESC 'Share Name'
+	EQUALITY caseIgnoreMatch
+	SYNTAX 1.3.6.1.4.1.1466.115.121.1.15 SINGLE-VALUE )
+
+attributetype ( 1.3.6.1.4.1.7165.2.1.42 NAME 'sambaOptionName'
+	DESC 'Option Name'
+	EQUALITY caseIgnoreMatch
+	SUBSTR caseIgnoreSubstringsMatch
+	SYNTAX 1.3.6.1.4.1.1466.115.121.1.15{256} )
+
+attributetype ( 1.3.6.1.4.1.7165.2.1.43 NAME 'sambaBoolOption'
+	DESC 'A boolean option'
+	EQUALITY booleanMatch
+	SYNTAX 1.3.6.1.4.1.1466.115.121.1.7 SINGLE-VALUE )
+
+attributetype ( 1.3.6.1.4.1.7165.2.1.44 NAME 'sambaIntegerOption'
+	DESC 'An integer option'
+	EQUALITY integerMatch
+	SYNTAX 1.3.6.1.4.1.1466.115.121.1.27 SINGLE-VALUE )
+
+attributetype ( 1.3.6.1.4.1.7165.2.1.45 NAME 'sambaStringOption'
+	DESC 'A string option'
+	EQUALITY caseExactIA5Match
+	SYNTAX 1.3.6.1.4.1.1466.115.121.1.26 SINGLE-VALUE )
+
+attributetype ( 1.3.6.1.4.1.7165.2.1.46 NAME 'sambaStringListOption'
+	DESC 'A string list option'
+	EQUALITY caseIgnoreMatch
+	SYNTAX 1.3.6.1.4.1.1466.115.121.1.15 )
+
+
+attributetype ( 1.3.6.1.4.1.7165.2.1.50 NAME 'sambaPrivName' 
+	SUP name )
+
+attributetype ( 1.3.6.1.4.1.7165.2.1.52 NAME 'sambaPrivilegeList'
+	DESC 'Privileges List'
+	EQUALITY caseIgnoreIA5Match
+	SYNTAX 1.3.6.1.4.1.1466.115.121.1.26{64} )
+
+attributetype ( 1.3.6.1.4.1.7165.2.1.53 NAME 'sambaTrustFlags'
+	DESC 'Trust Password Flags'
+	EQUALITY caseIgnoreIA5Match
+	SYNTAX 1.3.6.1.4.1.1466.115.121.1.26 )
+
 
 #######################################################################
 ##              objectClasses used by Samba 3.0 schema               ##
@@ -350,6 +414,14 @@
 	MAY  ( displayName $ description $ sambaSIDList ))
 
 ##
+## Trust password for trust relationships (any kind)
+##
+objectclass ( 1.3.6.1.4.1.7165.2.2.14 NAME 'sambaTrustPassword' SUP top STRUCTURAL
+	DESC 'Samba Trust Password'
+	MUST ( sambaDomainName $ sambaNTPassword $ sambaTrustFlags )
+	MAY ( sambaSID $ sambaPwdLastSet ))
+
+##
 ## Whole-of-domain info
 ##
 objectclass ( 1.3.6.1.4.1.7165.2.2.5 NAME 'sambaDomain' SUP top STRUCTURAL
@@ -359,7 +431,9 @@
 	MAY ( sambaNextRid $ sambaNextGroupRid $ sambaNextUserRid $
 	      sambaAlgorithmicRidBase ) )
 
+##
 ## used for idmap_ldap module
+##
 objectclass ( 1.3.6.1.4.1.7165.2.2.7 NAME 'sambaUnixIdPool' SUP top AUXILIARY
         DESC 'Pool for allocating UNIX uids/gids'
         MUST ( uidNumber $ gidNumber ) )
@@ -371,6 +445,27 @@
 	MAY ( uidNumber $ gidNumber ) )
 
 objectclass ( 1.3.6.1.4.1.7165.2.2.9 NAME 'sambaSidEntry' SUP top STRUCTURAL
-        DESC 'Structural Class for a SID'
-        MUST ( sambaSID ) )
+	DESC 'Structural Class for a SID'
+	MUST ( sambaSID ) )
+
+objectclass ( 1.3.6.1.4.1.7165.1.2.2.10 NAME 'sambaConfig' SUP top AUXILIARY
+	DESC 'Samba Configuration Section'
+	MAY ( description ) )
+
+objectclass ( 1.3.6.1.4.1.7165.2.2.11 NAME 'sambaShare' SUP top STRUCTURAL
+	DESC 'Samba Share Section'
+	MUST ( sambaShareName )
+	MAY ( description ) )
+
+objectclass ( 1.3.6.1.4.1.7165.2.2.12 NAME 'sambaConfigOption' SUP top STRUCTURAL
+	DESC 'Samba Configuration Option'
+	MUST ( sambaOptionName )
+	MAY ( sambaBoolOption $ sambaIntegerOption $ sambaStringOption $ 
+	      sambaStringListoption $ description ) )
+
+
+objectclass ( 1.3.6.1.4.1.7165.2.2.13 NAME 'sambaPrivilege' SUP top AUXILIARY
+	DESC 'Samba Privilege'
+	MUST ( sambaSID )
+	MAY ( sambaPrivilegeList ) )
 
